Shiny (Python) ML training example

neelasha23 opened the pull request:
Add Shiny example

<!-- readthedocs-preview ploomber-doc start -->
----
📚 Documentation preview 📚: https://ploomber-doc--105.org.readthedocs.build/en/105/

<!-- readthedocs-preview ploomber-doc end -->

edublancas commented on doc/_toc.yml:
```diff
@@ -29,6 +29,7 @@ parts:
       - file: apps/panel
       - file: apps/chainlit
       - file: apps/gradio
+      - file: apps/shiny-core
```
let's call this shiny, I'm unfamiliar with the "shiny core" term and other might be confused as well

neelasha23 replied:
[fixed](https://github.com/ploomber/doc/blob/6318fe9c144499deef0d0550c0626f73f6942aaf/doc/_toc.yml#L32)

edublancas commented on doc/apps/shiny-core.md:
```diff
@@ -0,0 +1,43 @@
+# Shiny Core
```
rename al references from "shiny core" to "shiny"

neelasha23 replied:
[fixed](https://github.com/ploomber/doc/blob/6318fe9c144499deef0d0550c0626f73f6942aaf/doc/apps/shiny.md?plain=1#L1)
Changed all references

edublancas commented on examples/docker/shiny-core/README.md:
```diff
@@ -0,0 +1,58 @@
+
+## Connecting to MongoDB Cloud
+
+For the purpose of this tutorial we will use the `MongoDB Atlas` cloud database. Let's see the steps for creating a cluster:
```
mention that this service allows you to host a database for free

neelasha23 replied:
[fixed](https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R10-R11)

edublancas commented on examples/docker/shiny-core/README.md:
```diff
@@ -0,0 +1,58 @@
+
+## Deploy application
+
+Create a zip file using `Dockerfile`, `app-core.py` and `requirements.txt` and deploy this file to Ploomber cloud.
```
add a link to the documentation that shows how to deploy shiny apps

neelasha23 replied:
[fixed](https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R40-R41)

edublancas commented on examples/docker/shiny-core/README.md:
````diff
@@ -0,0 +1,58 @@
+First create a virtual environment and install the required packages:
+
+```bash
+pip install tensorflow scikit-learn "pymongo[srv]"
````
so we need scikit-learn? I don't see it imported

neelasha23 replied:
It's there in train_one.py and train_two.py

edublancas commented on examples/docker/shiny-core/README.md:
````diff
@@ -0,0 +1,58 @@
+pip install tensorflow scikit-learn "pymongo[srv]"
+```
+
+Then run the training scripts parallely.:
````
```suggestion
Then run the training scripts parallely:
```

neelasha23 replied:
Sentenced replaced by : https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R59-R60

edublancas commented on examples/docker/shiny-core/README.md:
````diff
@@ -0,0 +1,58 @@
+Then run the training scripts parallely.:
+
+```bash
+python train_one.py & python train_two.py & wait
````
`&` is not available on windows, you can tell people to open two terminal windows and execute the commands on each one.

also, the scripts are pretty much the same, what changes between those two? It'd be better to have one script and just pass an argument:

```
python train.py --some-arg value
python train.py --some-arg another
```

neelasha23 replied:
[fixed](https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R59-R67)

edublancas commented on examples/docker/shiny-core/app-core.py:
```diff
@@ -0,0 +1,151 @@
+from __future__ import annotations
```
this is only needed for Python 2, I don't think it's used in this script

neelasha23 replied:
[Removed](https://github.com/ploomber/doc/pull/105/files#diff-c99010220192b232032cc417d9e26e799aa65a2dd2fbb5dd79b5595548293670R1-R3) from `app.py`

edublancas commented on examples/docker/shiny-core/app-core.py:
```diff
@@ -0,0 +1,151 @@
+THRESHOLD_LOW = 0.5
+THRESHOLD_LOW_COLOR = "rgb(193, 0, 0)"
+
+URI = "CONNECTION_STRING"
```
read this from and env variable

neelasha23 replied:
[fixed](https://github.com/ploomber/doc/pull/105/files#diff-c99010220192b232032cc417d9e26e799aa65a2dd2fbb5dd79b5595548293670R18-R19)

edublancas commented on examples/docker/shiny-core/README.md:
```diff
@@ -0,0 +1,58 @@
+
+![](./static/ip-access.png)
+
+Once you have completed the access rules setup, select the `CONNECT` button on the `Overview` page to get the connection details. Specifically, copy the connection string and replace the `URI` variable value in `app-core.py`, `train_one.py` and `train_two.py`
```
show how to set and environment variable and modify the scripts to read from there

neelasha23 replied:
[fixed](https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R42-R44)

[fixed](https://github.com/ploomber/doc/pull/105/files#diff-883b5ca3546bed48155f3709e49ebe5d5b987079fdc8b9c9d9f930a943512616R29)

edublancas commented on examples/docker/shiny-core/app-core.py:
```diff
@@ -0,0 +1,151 @@
+from __future__ import annotations
+
+import pandas as pd
```
rename this to app.py

neelasha23 replied:
Renamed

edublancas reviewed (changes requested):
please deploy the app so I can test it

neelasha23 commented:
Addressed the comments. App is deployed here: https://crimson-glade-4646.ploomberapp.io/
After setting the `MONGODB_CONNECTION_URI` env variable in local u can run the two scripts as mentioned in Readme: `python train.py --model model_1 --units 128` and `python train.py --model model_2 --units 12`
@edublancas

lfunderburk reviewed (commented):
Hi @neelasha23 

Took a look at the app - but the plot is not showing up
<img width="2508" alt="Screenshot 2024-02-02 at 4 00 51 PM" src="https://github.com/ploomber/doc/assets/13559004/eec75424-b057-4f73-8afa-c8f1180fff2b">

neelasha23 commented:
> Hi @neelasha23
> 
> Took a look at the app - but the plot is not showing up <img alt="Screenshot 2024-02-02 at 4 00 51 PM" width="2000" 

I have cleared the db because it's a realtime monitoring. you need to connect to ur mongodb account and run training scripts from local and view the plot in the dashboard. have added instructions in the README @lfunderburk

edublancas commented on examples/docker/shiny/train.py:
```diff
@@ -0,0 +1,83 @@
+        self.model_name = model_name
+
+    def on_epoch_end(self, epoch, logs=None):
+        my_collection.insert_one({'model': self.model_name, 'score': logs["accuracy"], 'timestamp': datetime.now()})
```
add a print statement here so users can compare what they see in the terminal with what they see in the app

neelasha23 replied:
[fixed](https://github.com/ploomber/doc/pull/105/files#diff-de90239477773a3e77cf5636a10f5db4b7b1bab1331462d3f5330d480173cf4cR35-R36)

edublancas commented:
the polling logic is off: I tried the app many times (both the ploomberapp.io one and locally) and I only got it one time to work. I debugged it by printing the value of `last_modified` and realized that even if I started training, the returned value was the same (I tried refreshing the browser after I started training).

I changed the `last_modified` function to this:

```python
def last_modified():
    n = my_collection.count_documents({})
    return n
```

and it worked better: as soon as I started training, the plot started updating. but I found another issue: the values in the plot did not match the values printed from the terminal and when starting to train another model, the plot didn't show a new line:


https://github.com/ploomber/doc/assets/989250/f9ba9706-86bb-4239-b8b2-794dcc14b485

neelasha23 commented:
I haven't been able to reproduce the timestamp not getting updated issue. Did you try with my account? I had selected an Indian server, so I'm doubting if that's causing a lag.
Can you please try once with a new account? If it still doesn't work I'll figure out what's happening
@edublancas 


https://github.com/ploomber/doc/assets/29096135/b1053d2b-a84b-497f-82d1-e51a86f88144

edublancas commented:
@neelasha23 the problem is most likely the `datetime.now()`, such function returns the timestamp in your local timezone.

but there's no need to reproduce it since I already provided a fix:

```python
def number_of_observations():
    n = my_collection.count_documents({})
    return n
```

this function will trigger an update as soon as there is a new element in the collection

neelasha23 commented:
Fixed the polling function: https://github.com/ploomber/doc/pull/105/files#diff-c99010220192b232032cc417d9e26e799aa65a2dd2fbb5dd79b5595548293670R33-R38

Although I did not completely understand why polling the timestamp column would be an issue. Even if datetime generates in local timezone it will still get accurately stored in MongoDB. And the polling function keeps checking this stored timestamp to detect if any new row has been added. You also mentioned that even after changing the polling function the timestamps in the plot are not matching the training ones , but I don't see that happening in my local. Which is why I feel maybe there's a lag between the training and the storage/retrieval process in your case? 
@edublancas 


https://github.com/ploomber/doc/assets/29096135/49d4b758-bd3e-4927-9046-1ba8e7b72fe5
